refactor: move build code to esm from commonjs

sgtcoolguy opened the pull request:
Related to #12739

sgtcoolguy committed:
refactor: move build code to esm from commonjs

build commented:
<!--
  0 failure: 
  2 warning:  :mag: Can't find ..., There is no linke...
  2 messages
  
  DangerID: danger-id-Danger;
-->


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  :mag: Can't find junit reports at `./junit.*.xml`, skipping generating JUnit Report.
  </td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>There is no linked JIRA ticket in the PR body. Please include the URL of the relevant JIRA ticket. If you need to, you may file a ticket on <a href="https://jira.appcelerator.org/secure/CreateIssue!default.jspa">JIRA</a></td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>:fist: The commits in this PR match our conventions! Feel free to Rebase and Merge this PR when ready.</td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>:thumbsup: Hey!, You deleted more code than you added. That's awesome!</td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against c6f05f7fd953965b72e8aad399fe994da74887dd
</p>

sgtcoolguy committed:
fix: make node happy?

ewanharris commented on android/titanium/libv8-services.mjs:
```diff
@@ -129,7 +131,7 @@ async function generateSnapshot(v8SnapshotHeaderFilePath, rollupFileContent) {
 
 	// Generate 'V8Snapshots.h' from template
-	const template = await util.promisify(ejs.renderFile)('V8Snapshots.h.ejs', { blobs }, {});
+	const template = await promisify(ejs.renderFile)('V8Snapshots.h.ejs', { blobs }, {});
```
Nit, we might be able to remove the promisify here, it appears `ejs.renderFile` can return a promise since [2.5.8](https://github.com/mde/ejs/blob/main/CHANGELOG.md#v258-2018-03-25)

ewanharris commented on build/scons.js:
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env node
+#!/opt/homebrew/bin/node
```
I think this should stay unchanged?
```suggestion
#!/usr/bin/env node
```

ewanharris commented on android/titanium/prebuild.mjs:
```diff
@@ -183,16 +185,14 @@ async function generateSourceCode() {
-			process.exit(1);
-		});
-}
+main()
```
These scripts will all be so much nicer when we can use top level await 😅

ewanharris reviewed (commented):
Only thing would be the shebang I think. What're the steps to landing this? Move all of the dev build scripts over to esm?

ewanharris commented on build/lib/builder.mjs:
```diff
@@ -1,17 +1,20 @@
+import path from 'path';
+import fs from 'fs-extra';
+import { rollup } from 'rollup';
+import babel from '@rollup/plugin-babel';
```
Oops, shoulda checked CI, this should be as below 
```suggestion
import { babel } from '@rollup/plugin-babel';
```

ewanharris committed:
build: correct babel import

ewanharris committed:
ci: use newer node version on Jenkins

The esm build scripts require nodejs/node/pull/35249 which landed in 12.20.0 and above

ewanharris committed:
test: move build test to esm

ewanharris commented:
Marked as ready for review accidentally in the app.

@hansemannn this is the scons build code which is less problematic imo, I remember reviewing this and what as changed worked ok. There was just no real path determined to getting it merged and it's not a real high priority either

cb1kenobi commented:
I approve this.

With that said, I don't love Node.js's .mjs extension and once we get the Titanium CLI moved to ES modules, we can probably set the package type to "module" and no longer need the .mjs extension. So, it's a race between the merge conflicts getting fixed in this PR (along with the nitpicks) and the Titanium CLI being massively refactored. Go!

hansemannn commented:
Okay, then let's fix the merge conflicts and go with it. I think the CLI should rather move to TypeScript instead of modular JS, but I'm not in the position to dictate that. From what I see in our projects, it's all TS and it's superb.

cb1kenobi commented:
Closing as we've already converted the SDK to ESM.
